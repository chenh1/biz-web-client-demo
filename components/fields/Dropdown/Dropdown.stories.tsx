import type { DropdownProps } from './Dropdown';
import React from 'react';
import { Dropdown } from './Dropdown';

type Args = {
  args: DropdownProps
}

const Template = (props: DropdownProps) => (
  <div className="pt-6">
    <Dropdown {...props}/>
  </div>
)

export default {
  component: Template
}

const options = [
  { value: 'default', label: 'Select an option', isPlaceholder: true },
  { value: 'novice', label: 'Novice' },
  { value: 'beginner', label: 'Beginner' },
  { value: 'intermediate', label: 'Intermediate' },
  { value: 'advanced', label: 'Advanced' }
]

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const DropdownHyperRed: Args = {
  args: {
    highlightColor: "hyper-red",
    options,
  },
};

export const DropdownUltraPurple: Args = {
  args: {
    highlightColor: "ultra-purple",
    options,
  },
};

export const DropdownForceYellow: Args = {
  args: {
    highlightColor: "force-yellow",
    options,
  },
};

export const DropdownApexBlue: Args = {
  args: {
    highlightColor: "apex-blue",
    options,
  },
};

export const DropdownHyperRedError: Args = {
  args: {
    highlightColor: "hyper-red",
    options,
    isError: true
  },
};

export const DropdownHyperRedConfirmed: Args = {
  args: {
    highlightColor: "hyper-red",
    options,
    isConfirmed: true
  },
};