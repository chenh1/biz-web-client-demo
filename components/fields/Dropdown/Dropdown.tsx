"use client";

import { getTextColorClass, getBorderColorClass } from '../../constants/colors';
import type { CoreColors } from '../../constants/colors';
import React, { useMemo, useState, ChangeEvent } from 'react';
import { CLICK_CLASS, SECTION_VIEW_CLASS } from '../../constants/tagging';

interface Option {
  value: string;
  label: string;
  isPlaceholder?: boolean;
}

export type DropdownProps = {
  onChange?: (event: ChangeEvent<HTMLSelectElement>) => void;
  useLabelPlaceholder?: boolean;
  options: Option[];
  disabled?: boolean;
  iconAlign?: 'left' | 'right';
  isError?: boolean;
  isConfirmed?: boolean;
  isInlineBlock?: boolean;
  name?: string;
  highlightColor?: CoreColors;
  fullWidth?: boolean;
  gtmClick?: boolean;
  gtmSectionView?: boolean;
  dataGtmId?: string;
}

export const Dropdown: React.FC<DropdownProps> = ({
  onChange = () => {},
  useLabelPlaceholder,
  options,
  disabled,
  iconAlign,
  isError,
  isConfirmed,
  isInlineBlock,
  name,
  highlightColor = 'hyper-red',
  fullWidth,
  gtmClick,
  gtmSectionView,
  dataGtmId
}) => {
  const [isFocused, setIsFocused] = useState<boolean>(false);

  const computedTextColorClass = useMemo(() => getTextColorClass(highlightColor), [highlightColor]);
  const computedBorderColorClass = useMemo(() => getBorderColorClass(highlightColor), [highlightColor]);
  const computedWidthClass = useMemo(() => (fullWidth ? 'w-full' : ''), [fullWidth]);

  const placeholder = options.find(option => option.isPlaceholder)?.label || 'Select an option';

  return (
    <div className={`relative ${isInlineBlock ? 'inline-block' : ''}`}>
      {useLabelPlaceholder &&
        <label 
          data-gtm-id={dataGtmId}
          htmlFor={name} 
          className={`${gtmClick ? CLICK_CLASS : ''} ${gtmSectionView ? SECTION_VIEW_CLASS : ''} pointer-events-none font-body duration-300 absolute block text-md font-medium ${computedTextColorClass} -top-5 text-sm`}
        >
          {placeholder}
        </label>
      }
      <select 
        data-gtm-id={dataGtmId}
        disabled={disabled}
        onChange={onChange} 
        onFocus={() => setIsFocused(true)} 
        onBlur={() => setIsFocused(false)} 
        name={name} 
        className={`${gtmClick ? CLICK_CLASS : ''} ${gtmSectionView ? SECTION_VIEW_CLASS : ''} max-w-full ${computedWidthClass} text-night-black dark:text-max-white font-body bg-transparent border-b-2 focus:outline-none ${isFocused ? computedBorderColorClass : 'border-night-black dark:border-max-white'}`}
      >
        {options.map((option, i) => (
          <option disabled={option.isPlaceholder} key={i} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
      {isError && <div className={`pointer-events-none absolute ${iconAlign === 'right' ? 'right-0' : 'left-0'} top-0 -translate-x-full`}>&#10060;</div>}
      {isConfirmed && <div className={`pointer-events-none absolute ${iconAlign === 'right' ? 'right-0' : 'left-0'} top-0 -translate-x-full`}>&#x2705;</div>}
    </div>
  );
};
