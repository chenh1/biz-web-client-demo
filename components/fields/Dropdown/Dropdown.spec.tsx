import { Dropdown } from "./Dropdown";
import { fireEvent, render } from '@testing-library/react';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

const options = [
  { value: 'default', label: 'Select an option', isPlaceholder: true },
  { value: 'novice', label: 'Novice' },
  { value: 'beginner', label: 'Beginner' },
  { value: 'intermediate', label: 'Intermediate' },
  { value: 'advanced', label: 'Advanced' }
]

it('renders a dropdown field with options', async () => {
  const { asFragment } = render(
    <Dropdown options={options} />
  );

  expect(asFragment()).toMatchSnapshot();
});

it('renders a dropdown field with an error state', async () => {
  const { asFragment } = render(
    <Dropdown options={options} isError={true} />
  );

  expect(asFragment()).toMatchSnapshot();
});


it('renders a dropdown field with a confirm state', async () => {
  const { asFragment } = render(
    <Dropdown options={options} isConfirmed={true} />
  );

  expect(asFragment()).toMatchSnapshot();
});

it('renders a dropdown field with a confirm AND error state if both are available', async () => {
  const { asFragment } = render(
    <Dropdown options={options} isConfirmed={true} isError={true}/>
  );

  expect(asFragment()).toMatchSnapshot();
});
