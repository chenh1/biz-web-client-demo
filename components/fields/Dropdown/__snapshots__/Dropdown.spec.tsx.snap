// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a dropdown field with a confirm AND error state if both are available 1`] = `
<DocumentFragment>
  <div
    class="relative "
  >
    <select
      class="  max-w-full  text-night-black dark:text-max-white font-body bg-transparent border-b-2 focus:outline-none border-night-black dark:border-max-white"
    >
      <option
        disabled=""
        value="default"
      >
        Select an option
      </option>
      <option
        value="novice"
      >
        Novice
      </option>
      <option
        value="beginner"
      >
        Beginner
      </option>
      <option
        value="intermediate"
      >
        Intermediate
      </option>
      <option
        value="advanced"
      >
        Advanced
      </option>
    </select>
    <div
      class="pointer-events-none absolute left-0 top-0 -translate-x-full"
    >
      ❌
    </div>
    <div
      class="pointer-events-none absolute left-0 top-0 -translate-x-full"
    >
      ✅
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders a dropdown field with a confirm state 1`] = `
<DocumentFragment>
  <div
    class="relative "
  >
    <select
      class="  max-w-full  text-night-black dark:text-max-white font-body bg-transparent border-b-2 focus:outline-none border-night-black dark:border-max-white"
    >
      <option
        disabled=""
        value="default"
      >
        Select an option
      </option>
      <option
        value="novice"
      >
        Novice
      </option>
      <option
        value="beginner"
      >
        Beginner
      </option>
      <option
        value="intermediate"
      >
        Intermediate
      </option>
      <option
        value="advanced"
      >
        Advanced
      </option>
    </select>
    <div
      class="pointer-events-none absolute left-0 top-0 -translate-x-full"
    >
      ✅
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders a dropdown field with an error state 1`] = `
<DocumentFragment>
  <div
    class="relative "
  >
    <select
      class="  max-w-full  text-night-black dark:text-max-white font-body bg-transparent border-b-2 focus:outline-none border-night-black dark:border-max-white"
    >
      <option
        disabled=""
        value="default"
      >
        Select an option
      </option>
      <option
        value="novice"
      >
        Novice
      </option>
      <option
        value="beginner"
      >
        Beginner
      </option>
      <option
        value="intermediate"
      >
        Intermediate
      </option>
      <option
        value="advanced"
      >
        Advanced
      </option>
    </select>
    <div
      class="pointer-events-none absolute left-0 top-0 -translate-x-full"
    >
      ❌
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders a dropdown field with options 1`] = `
<DocumentFragment>
  <div
    class="relative "
  >
    <select
      class="  max-w-full  text-night-black dark:text-max-white font-body bg-transparent border-b-2 focus:outline-none border-night-black dark:border-max-white"
    >
      <option
        disabled=""
        value="default"
      >
        Select an option
      </option>
      <option
        value="novice"
      >
        Novice
      </option>
      <option
        value="beginner"
      >
        Beginner
      </option>
      <option
        value="intermediate"
      >
        Intermediate
      </option>
      <option
        value="advanced"
      >
        Advanced
      </option>
    </select>
  </div>
</DocumentFragment>
`;
